llvm/docs/LangRef.rst

```diff
@@ -3394,17 +3394,41 @@ Floating-Point Environment
 The default LLVM floating-point environment assumes that traps are disabled and
 status flags are not observable. Therefore, floating-point math operations do
 not have side effects and may be speculated freely. Results assume the
-round-to-nearest rounding mode.
+round-to-nearest rounding mode, and subnormals are assumed to be preserved.
+Running default LLVM code in an environment where these assumptions are not met
+can lead to undefined behavior.
+
+The representation bits of a floating-point value do not mutate arbitrarily; if
+there is no floating-point operation being performed, the NaN payload (if any)
+is preserved.
```

Comment:
floating point bits *do* mutate arbitrarily on 32-bit x86:
here, all I do is bitcast `0x7f800001` to `float`, return it from a function, then bitcast back to an integer -- that changes it to `0x7fc00001`.
https://clang.godbolt.org/z/sjaq55q3G

Reply:
Yes, that's a bug in the x86-32 backend / calling convention.

Given that this bug is hard-coded in the calling convention, I wonder if it should be called out here. But the main point if this sentence is that LLVM itself will not do anything that would mutate FP values. In fact this bug here leads to miscompilations because LLVM itself *assumes* that FP values don't change -- see https://github.com/llvm/llvm-project/issues/44218.

Reply:
I do think we should at least mention the major ways in which LLVM backends violate the spec:
- x86-32 with SSE2 enabled may implicitly convert floating-point values returned from a function to x86_fp80 and back; such conversions may signal a floating-point exception, or modify the value.
- x86-32 without SSE2 may do such conversions just about _anywhere_.
- older MIPS versions use the opposite polarity for the sNaN vs qNaN bit, and LLVM does not correctly represent this (this is at least potentially fixable, unlike the above).

Reply:
I added a paragraph with backend problems and linked to relevant issues. Is there an issue for the "x86-32 return value problem"?

Reply:
I'm not aware of one in LLVM, only in rust, so I filed a new one, #66803.